aeron-client/src/test/c/aeron_image_fragment_assembler_test.cpp

```diff
@@ -0,0 +1,348 @@
+    aeron_image_fragment_assembler_t *m_assembler = nullptr;
+};
+
+TEST_F(ImageFragmentAssemblerTest, shouldPassThroughUnfragmentedMessage)
+{
+    size_t fragmentLength = 158;
+    fillFrame(AERON_DATA_HEADER_UNFRAGMENTED, 0, fragmentLength, 0);
+    bool isCalled = false;
+    auto handler = [&](const uint8_t *buffer, size_t length, aeron_header_t *header)
+    {
+        isCalled = true;
+        EXPECT_EQ(length, fragmentLength);
+        aeron_header_values_t header_values;
+        EXPECT_EQ(0, aeron_header_values(header, &header_values));
+        EXPECT_EQ(SESSION_ID, header_values.frame.session_id);
+        verifyPayload(buffer, length);
+    };
+
+    handle_fragment(handler, fragmentLength);
+    EXPECT_TRUE(isCalled);
+}
+
+TEST_F(ImageFragmentAssemblerTest, shouldReassembleFromTwoFragments)
+{
+    size_t fragmentLength = MTU_LENGTH - AERON_DATA_HEADER_LENGTH;
+    int32_t initialTermId = 420;
+    size_t position_bits_to_shift = 20;
+    int8_t version = 42;
+    int16_t type = AERON_HDR_TYPE_ERR;
+    int32_t sessionId = 18;
+    int32_t streamId = 31;
+    int32_t termId = initialTermId + 5;
+    int64_t reservedValue = -47923742793L;
+
+    int32_t termOffset = 0;
+    bool isCalled = false;
+    auto handler = [&](const uint8_t *buffer, size_t length, aeron_header_t *header)
+    {
+        isCalled = true;
+        EXPECT_EQ(length, fragmentLength * 2);
+        aeron_header_values_t header_values;
+        EXPECT_EQ(0, aeron_header_values(header, &header_values));
+        EXPECT_EQ(initialTermId, header_values.initial_term_id);
+        EXPECT_EQ(position_bits_to_shift, header_values.position_bits_to_shift);
+        EXPECT_EQ(MTU_LENGTH * 2, header_values.frame.frame_length);
+        EXPECT_EQ((uint8_t)(AERON_DATA_HEADER_END_FLAG | AERON_DATA_HEADER_BEGIN_FLAG), header_values.frame.flags);
+        EXPECT_EQ(version, header_values.frame.version);
+        EXPECT_EQ(type, header_values.frame.type);
+        EXPECT_EQ(0, header_values.frame.term_offset);
+        EXPECT_EQ(sessionId, header_values.frame.session_id);
+        EXPECT_EQ(streamId, header_values.frame.stream_id);
+        EXPECT_EQ(termId, header_values.frame.term_id);
+        EXPECT_EQ(reservedValue, header_values.frame.reserved_value);
+        verifyPayload(buffer, length);
+    };
+
+    m_header.initial_term_id = initialTermId;
+    m_header.position_bits_to_shift = position_bits_to_shift;
+    fillFrame(
+        version,
+        (uint8_t)(AERON_DATA_HEADER_BEGIN_FLAG | AERON_DATA_HEADER_EOS_FLAG),
+        type,
+        termOffset,
+        sessionId,
+        streamId,
+        termId,
+        reservedValue,
+        fragmentLength,
+        0);
+    handle_fragment(handler, fragmentLength);
+    EXPECT_FALSE(isCalled);
+
+    termOffset += MTU_LENGTH;
+    m_header.initial_term_id = -1;
+    m_header.position_bits_to_shift = -7;
+    fillFrame(
+        AERON_FRAME_HEADER_VERSION,
+        AERON_DATA_HEADER_END_FLAG,
+        AERON_HDR_TYPE_EXT,
```

Comment:
This will require changing the definition of `AERON_HDR_TYPE_EXT` to be `(-1)` or `((int16_t)0xFFFF)`, so that the type cast it explicit and there is no implicit overflow.  Fails with warnings as errors on GCC.

Reply:
Done.